### ISSUE TITLE
Preservation of receive buffer causes underflow

### ISSUE THREAD

**Issue by vorner** [jbaublitz/neli#45]
Sometimes, I get a panic like this:

```
thread 'main' panicked at 'attempt to subtract with overflow', /home/vorner/.cargo/git/checkouts/neli-94d2bc1fa408b168/ad0cb18/src/nl.rs:104:31
stack backtrace:
   0:     0x5555555810a4 - backtrace::backtrace::libunwind::trace::h7eb4b43b611b7867
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/libunwind.rs:88
   1:     0x5555555810a4 - backtrace::backtrace::trace_unsynchronized::hd915225b960621fe
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.37/src/backtrace/mod.rs:66
   2:     0x5555555810a4 - std::sys_common::backtrace::_print_fmt::h03e09238238aa7ea
                               at src/libstd/sys_common/backtrace.rs:77
   3:     0x5555555810a4 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hd2bc77662b898564
                               at src/libstd/sys_common/backtrace.rs:61
   4:     0x5555555994fc - core::fmt::write::h8ebd3dea7d60af99
                               at src/libcore/fmt/mod.rs:1030
   5:     0x55555557f687 - std::io::Write::write_fmt::hcc91da7d626d8d3c
                               at src/libstd/io/mod.rs:1412
   6:     0x555555583425 - std::sys_common::backtrace::_print::h71d790ff242dd25b
                               at src/libstd/sys_common/backtrace.rs:65
   7:     0x555555583425 - std::sys_common::backtrace::print::h0598a41f9f73dcfd
                               at src/libstd/sys_common/backtrace.rs:50
   8:     0x555555583425 - std::panicking::default_hook::{{closure}}::h402ca543df8941ce
                               at src/libstd/panicking.rs:200
   9:     0x555555583116 - std::panicking::default_hook::h654c705cfbcaa6d8
                               at src/libstd/panicking.rs:214
  10:     0x555555583a75 - std::panicking::rust_panic_with_hook::h56efa3c9265fc6c3
                               at src/libstd/panicking.rs:477
  11:     0x555555583612 - std::panicking::continue_panic_fmt::h21f1f952a36d712d
                               at src/libstd/panicking.rs:384
  12:     0x555555583506 - rust_begin_unwind
                               at src/libstd/panicking.rs:311
  13:     0x555555596eca - core::panicking::panic_fmt::hb830357db65719c4
                               at src/libcore/panicking.rs:85
  14:     0x555555596e09 - core::panicking::panic::hb2537515f34024e6
                               at src/libcore/panicking.rs:49
  15:     0x55555556222f - <neli::nl::Nlmsghdr<T,P> as neli::Nl>::deserialize::h41f512b50efc9532
                               at /home/vorner/.cargo/git/checkouts/neli-94d2bc1fa408b168/ad0cb18/src/nl.rs:104
```

I've discovered this is caused by preserving the self.buffer in socket after a receive of an error message ‒ and it seems the kernel sends something after the message, or I don't parse it fully correctly and there are some attributes I don't use.

It seems the next run of the recv_nl tries to reuse the buffer, or its leftover and as this is not the start of the message, it fails miserably.

I have few questions:
* Why is the buffer even preserved? What I know, netlink is not like TCP, it's message-based. If a buffer is too short, it gets truncated, and you never get multiple messages through one recv. Or is there some other reason for that?
* Should it panic/do such things if the message is garbage? Shouldn't it just return an error?

**Comment by jbaublitz** [comment]
Please see #15. There are cases where netlink will return multiple messages. There is an example in that issue that the reporter provided showing that multiple messages can correspond to a single read.

As for the panic, you're correct. That should not happen. I'll take a look at using `checked_sub` instead. I had another contributor add that elsewhere and I intend to standardize on that.

**Comment by vorner** [comment]
I'm really surprised that the kernel can send multiple messages in one read, I did think it always corresponded 1:1 with reads/writes.

Anyway, I still have a gut feeling what I have here is some kind of a leftover from the previous message or unparsed garbage, not another message. For one, I'm expecting just one. For another, when I start examining the variables in gdb, the amount of data in the buffer is smallish, so it definitely wasn't truncated ‒ but the line where the underflow happens would suggest there isn't enough data in the message to make sense. So, even if the code uses `checked_sub`, what will it do at that place?

If you want to have a more concrete data in hand, I can reproduce it with this code: https://github.com/vorner/nelilog/commit/6069967783b3af055795e4db3fbdc4b1d4330132. If I run it as an unprivileged user, I get a permission denied error first. Then the panic. Indeed, that code is more of an experiment than something „nice“ or even correct, but I probably shouldn't be able to destroy neli's internal data through the API.

**Comment by skrap** [comment]
> I'm really surprised that the kernel can send multiple messages in one read, I did think it always corresponded 1:1 with reads/writes.

It makes sense from a performance standpoint.  If the kernel can transmit multiple messages in one syscall, why wouldn't it?
I have personally witnessed this with the rtnetlink subsystem, so I think it's best to assume that it could happen in any subsystem.

**Comment by jbaublitz** [comment]
@vorner I agree that there shouldn't be a panic. I'm going to put up a PR to switch over to returning an error if the `usize` value drops below zero. However, the panic you've highlighted here indicates you are incorrectly parsing the data. the length of the entire packet can never be less than the size of the netlink header. I agree that this should be an informative error instead of what you're getting, but I'd recommend looking at the binary data and see if you are indeed specifying the appropriate types for what's being returned from the kernel.

**Comment by vorner** [comment]
Yes, I know, I was parsing it as the wrong type.

But optimally, parsing one message as the wrong type should not damage the socket to the state it won't be possible to consume other messages ‒ if I understand correctly, netlink messages are length prefixed. If I parse one incorrectly, it should still be possible to eat the right amount of data (either give up parsing the wrong type when the length runs out, or consume the extra).

**Comment by jbaublitz** [comment]
Thank you for your patience. I've been thinking on this a bit. I think providing some way to indicate that a message has failed parsing and moving on to the next one or allowing a re-parse is a great idea. I hadn't considered how to do that so I wanted to take some time to figure out what a reasonable implementation might be. Currently, this is what I propose:

1. Changes to `buffering` crate to support the position of both the beginning of the current message and the beginning of the next message. This would allow fast forwarding and discarding the current message or rewinding and retrying the message parsing.
2. Given that this check really only can be done at the message level, this would involve some special casing in the `Nlmsghdr` struct. This would allow storing the necessary information for replay and fast forwarding the buffer at the beginning of the `deserialize()` method. If a method below returns a `DeError` or the entire message is not consumed at the end of deserialization, return some error indicating that the message is not fully parsed and allow the user to handle either fast forwarding or rewinding through the `buffering` crate at the lowest level.
3. High level API considerations should probably be allowing the user to specify the default behavior when a message fails to parse. Either rewind or fast forward by default and also expose an API to override this if necessary.

How does this sound? I'm not set on how I've outlined this yet so I'd be happy to hear any suggestions you have.

**Comment by vorner** [comment]
To me it sounds reasonable, and probably solves part of #50, allowing to re-parse it as something different. So the error could contain optional u16 as the type and have methods to rewind (and then parse again as another type) and skip.

An alternative is to look for what others do. I've seen many use the bytes crate. That one has a buffer type that can be very cheaply split into smaller (owned) parts:
https://docs.rs/bytes/0.4.12/bytes/struct.Bytes.html (it implements the Buf trait, which allows reading basic types from it ergonomically).

That way, the receive buffer could be split into an iterator of `(u16, Bytes)` by the lengths. Then, the payload would be parsed from each `Bytes` buffer, which would already know its range. When parsing attributes and sub-attributes, the same way could be used uniformly ‒ splitting the `Bytes` into smaller chunks. That would solve this issue, #50 and #52 in one go. The disadvantage is obviously big API change and probably quite a lot of work to do it.

**Comment by jbaublitz** [comment]
This should be resolved in branch `v0.5.0-dev`. Closing.
